### PR TITLE
Move Qowaiv.DomainModel.Events.EventCollection to Qowaiv.Collections.ImmutableCollection

### PR DESCRIPTION
**Pull request by Corniel**
As this functionality should not be limited to building a collection of events (with conditional statements) I moved and renamed `Qowaiv.DomainModel.Events.EventCollection` to `Qowaiv.Collections.ImmutableCollection`.

I also made the `EventBuffer` immutable using the `ImmutableCollection` under the hood.

### CODE REVIEW

**Review by pmdevers — APPROVED** [review]

File: src/Qowaiv.DomainModel/Collections/ImmutableCollection.cs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+namespace Qowaiv.DomainModel.Collections
+{
+    /// <summary>Represents an immutable collection.</summary>
+    /// <remarks>
+    /// As a design choice, adding null is ignored. Also <see cref="IEnumerable"/>s
+    /// are added as collections, execpt for <see cref="string"/>.
```
Comment (LauraXiulan) [inline src/Qowaiv.DomainModel/Collections/ImmutableCollection.cs:12]: except

File: src/Qowaiv.DomainModel/Dynamic/DynamicEventDispatcher.cs [hunk 2]
```diff
@@ -85,6 +80,16 @@ private bool When(object @event, out object result)
             return lookup;
         }
 
+        /// <summary>Uses refelection to resolve `When` methods to use,
```
Comment (LauraXiulan) [inline src/Qowaiv.DomainModel/Dynamic/DynamicEventDispatcher.cs:83]: reflection

File: src/Qowaiv.DomainModel/EventBuffer.cs [hunk 3]
```diff
@@ -99,45 +102,27 @@ public EventBuffer(TId aggregateId, int version)
         /// The event to add.
         /// </param>
         public EventBuffer<TId> Add(object @event)
-        {
-            Guard.NotNull(@event, nameof(@event));
-            buffer.Add(@event);
-            return this;
-        }
+            => new(AggregateId, offset, CommittedVersion, buffer.Add<object>(@event));
```
Comment (LauraXiulan) [inline src/Qowaiv.DomainModel/EventBuffer.cs:105]: No guard needed anymore?

File: src/Qowaiv.DomainModel/EventBuffer.cs [hunk 4]
```diff
@@ -99,45 +102,27 @@ public EventBuffer(TId aggregateId, int version)
...
 
         /// <summary>Adds  events to the event buffer.</summary>
```
Comment (LauraXiulan) [inline src/Qowaiv.DomainModel/EventBuffer.cs:107]: double space

File: src/Qowaiv.DomainModel/EventBuffer.cs [hunk 5]
```diff
@@ -99,45 +102,27 @@ public EventBuffer(TId aggregateId, int version)
...
         public EventBuffer<TId> ClearCommitted()
-        {
-            var delta = CommittedVersion - offset;
-            buffer.RemoveRange(0, delta);
-            offset += delta;
-            return this;
-        }
+            => new(
+                aggregateId: AggregateId,
+                offset: CommittedVersion,
```
Comment (LauraXiulan) [inline src/Qowaiv.DomainModel/EventBuffer.cs:123]: is offset always the same as committedVersion?

File: src/Qowaiv.DomainModel/EventBuffer.cs (hunk 3 above)
Comment (Corniel) [reply src/Qowaiv.DomainModel/EventBuffer.cs:105]: No, a null event is now ignored.

File: src/Qowaiv.DomainModel/EventBuffer.cs (hunk 5 above)
Comment (Corniel) [reply src/Qowaiv.DomainModel/EventBuffer.cs:123]: For ClearCommitted that is the case.

**Pull request merged**